linode/manager
# fix: [M3-7796] - Correct Default VPC Scope and Adjust PAT Creation Behavior

## Description 📝
There is an issue with the current Personal Access Token (PAT) creation process. When using the "Select All" option, it incorrectly selects the VPC scope even when it is disabled. This behavior could lead to unintentional granting of access and needs to be corrected to improve the usability and security of the platform.

## Changes  🔄
- Adds a check and early return in `handleSelectAllScopes` function to explicitly handle the VPC case where `Read Only` is checked, which should instead be `None` for that scope.
- Modifies `allScopesAreTheSame` util to optionally pass in excluded scopes, allowing the correct "Select All" radio button to be checked when some scopes have a different default.
- Adds test coverage to `CreateAPITokenDrawer.test.tsx` and APIToken `utils.test.ts`.

## Preview 📷
| Before  | After   |
| ------- | ------- |
| ![Screenshot 2024-02-20 at 4 32 11 PM](https://github.com/linode/manager/assets/114685994/a278fe62-13c7-48ac-8279-48926841ccfb) | ![localhost_3000_profile_tokens (4)](https://github.com/linode/manager/assets/114685994/ec07c5e9-2139-4abc-8bb4-ea6b10aa78dc) |
| ![Screenshot 2024-02-20 at 4 32 50 PM](https://github.com/linode/manager/assets/114685994/1bc7ae27-8648-4aee-8703-f212e4a2d051) | ![Screenshot 2024-02-23 at 9 17 41 AM](https://github.com/linode/manager/assets/114685994/c05063ec-3c50-4b23-a468-cfa7cfd0715e) |

## How to test 🧪

### Prerequisites
(How to setup test environment)
- Check out this PR and `yarn dev`.

### Reproduction steps
(How to reproduce the issue, if applicable)
- Go to https://cloud.linode.com/profile/tokens.
- Click the "Select All" read only option and observe that the disabled VPC read only radio button is selected.
- Observe that when the PAT is created, its VPC scope will be Read Only. (This can be seen either in the request body or the View PAT drawer.)

### Verification steps
(How to verify changes)
- Go to https://localhost:3000/profile/tokens.
- Click the "Select All" read only option and verify that radio button is checked even though not all the scopes in that column are read only.
- Verify that the disabled VPC read only radio button is *not* selected. Instead, `None` remains selected.
- Verify that when the PAT is created, its VPC scope will be None. This can be seen either in the request body (the VPC scope will not be included) or the View PAT drawer.

## As an Author I have considered 🤔

*Check all that apply*
